### DISCUSSION TITLE
Some API are deprecated but no replacement was found

### DISCUSSION THREAD

**Discussion by hcxmj** [webpack/webpack#15945] [General]
when upgrading webpack to version 5.73.0 and webpack-dev-server to version 4.9.2. For example: 1、"devServer.use is not a function"，2、Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'clientLogLevel' / 'contentBase' / 'publicPath' / 'quiet' / 'watchOptions'.I hope to find some way to replace but no documentation.

**Comment by snitin315** [comment]
Please read the migration guide for webpack-dev-server v4 here:

https://github.com/webpack/webpack-dev-server/blob/master/migration-v4.md#migration-guide

**Comment by hcxmj** [comment]
> Please read the migration guide for webpack-dev-server v4 here:

Oh thanks a lot, I've been searching the webpack documentation and haven't been able to find the answer

**Reply by alexander-akait** [reply]
If you need a help, please provide a configuration

**Reply by hcxmj** [reply]
It prompts me ‘TypeError: devServer.use is not a function’. And did not find an attribute that can replace ‘quiet’ API，ValidationError: ' Invalid options object. Dev Server has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'quiet''

function addMiddleware(devServer) {
  var proxy = require(paths.appPackageJson).proxy;
  devServer.use(historyApiFallback({
    disableDotRule: true,
    htmlAcceptHeaders: proxy ?
      ['text/html'] :
      ['text/html', '*/*']
  }));
  if (proxy) {
    if (typeof proxy !== 'string') {
      console.log(chalk.red('When specified, "proxy" in package.json must be a string.'));
      console.log(chalk.red('Instead, the type of "proxy" was "' + typeof proxy + '".'));
      console.log(chalk.red('Either remove "proxy" from package.json, or make it a string.'));
      process.exit(1);
    }
    var mayProxy = /^(?!\/(index\.html$|.*\.hot-update\.json$|sockjs-node\/)).*$/;
    var hpm = httpProxyMiddleware(pathname => mayProxy.test(pathname), {
      target: proxy,
      logLevel: 'silent',
      onProxyReq: function(proxyReq) {
        if (proxyReq.getHeader('origin')) {
          proxyReq.setHeader('origin', proxy);
        }
      },
      onError: onProxyError(proxy),
      secure: false,
      changeOrigin: true,
      ws: true,
      xfwd: true
    });
    devServer.use(mayProxy, hpm);
    devServer.listeningApp.on('upgrade', hpm.upgrade);
  }
  devServer.use(devServer.middleware);
}

  var devServer = new WebpackDevServer({
    compress: false,
    // clientLogLevel: 'none',
    client: {
      logging: "none",
    },
    static: {
      directory: paths.appStatic, // --- contentBase: paths.appPublic,
      watch: { // watchOptions: {---
        ignored: '/node_modules/'
      }
    },
   quiet: true, // -----this API
    hot: false,
    devMiddleware: {
      publicPath: config.output.publicPath, // ---publicPath: config.output.publicPath,
    },
    https: protocol === "https",
    host: host,
    proxy: {
      '/sso': 'http://apm.qiyi.domain',
      '/apmconfig': 'http://apm.qiyi.domain',
      '/wsgi': 'http://apm.qiyi.domain',
      '/rest': 'http://apm.qiyi.domain',
      '/druid': 'http://apm.qiyi.domain',
      '/kylin': 'http://apm.qiyi.domain',
      '/apm': 'http://apm.qiyi.domain',
      '/qmas': 'http://qmasapi.qiyi.domain',
      '/api/GetPassword': 'http://itpwd.qiyi.domain',
      '/urls/yourls-api.php': 'http://arch.qiyi.domain',
      '/api/open': 'http://hubble.qiyi.domain/',
      '/api/alarm': 'http://hubble.qiyi.domain/',
    }
  }, compiler);
  addMiddleware(devServer);

**Reply by alexander-akait** [reply]
Change `devServer.use` and `devServer.listeningApp.YOUR_METHOD` on `devServer.app.use` and `quiet` was removed

**Reply by hcxmj** [reply]
ok, is there a more systematic and comprehensive document to see the changes of each version.Because I just used devServer.app.use and it still prompts a new error.(TypeError: Cannot read properties of undefined (reading 'use'))
`  devServer.app.use(historyApiFallback({
                ^

TypeError: Cannot read properties of undefined (reading 'use')
    at addMiddleware (D:\work\apmweb\scripts\start.js:231:17)
    at runDevServer (D:\work\apmweb\scripts\start.js:372:3)
    at run (D:\work\apmweb\scripts\start.js:394:3)
    at D:\work\apmweb\scripts\start.js:401:5
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`

**Reply by hcxmj** [reply]
Probably my upgrade version span is relatively large.
webpack version from `1.14.0` to '^5.73.0'.
webpack-dev-server version from 1.16.2 to '^4.9.2' and webpack-cli version is '^4.9.2'.
